### PR TITLE
Origin/darreldonald/log analytics data provider

### PR DESCRIPTION
**Pull request by DarrelDonald**
I have added a data provider for K8SE Log Analytics. 

This data provider is specific to K8SE.
&nbsp;&nbsp;&nbsp;&nbsp; A generic Log Analytics data provider would require users to pass the workspace ID
&nbsp;&nbsp;&nbsp;&nbsp; This workspace is in AME tenant and wouldn't be able to use the same app id/secret as workspaces in Microsoft tenant.

The method to call in the detector code to use this data provider is `dp.K8SELogAnalytics.ExecuteQueryAsync` 
&nbsp;&nbsp;&nbsp;&nbsp;It takes a log analytics query as a string for its input and returns the result as a data table.

Example:
&nbsp;&nbsp;&nbsp;&nbsp;```var tableTask = dp.K8SELogAnalytics.ExecuteQueryAsync("union * | take 5");```

### CODE REVIEW

File: src/Diagnostics.RuntimeHost/Diagnostics.RuntimeHost.csproj [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk.Web">
 
   <PropertyGroup>
-    <TargetFramework>netcoreapp2.2</TargetFramework>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
```
Comment (rekhaswaminathan) [inline src/Diagnostics.RuntimeHost/Diagnostics.RuntimeHost.csproj:4]: Was this change intended?

**Comment by rekhaswaminathan** [comment]
Please consider writing a simple HealthCheck method that validates the connectivity to LogAnalytics endpoint. This is helpful in catching any exceptions before the impact reaches production and is also part of our devops process.

Eg: 
https://github.com/Azure/Azure-AppServices-Diagnostics/blob/524926bcb3607cba6ab0548d0317cce39738b430/src/Diagnostics.DataProviders/DataProviders/KustoDataProvider.cs#L196

**Comment by rekhaswaminathan** [comment]
Can we rename K8SELogAnalytics to something more meaningful? The term 'K8SE' might not be clear for someone new/not familiar with the system.

File: src/Diagnostics.DataProviders/Interfaces/ILogAnalyticsClient.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Diagnostics.DataProviders.Interfaces
+{
+    public interface ILogAnalyticsClient
+    {
+    }
```
Comment (rekhaswaminathan) [inline src/Diagnostics.DataProviders/Interfaces/ILogAnalyticsClient.cs:9]: This class is empty. Declare the method here- `ExecuteQueryAsync(string query)`

File: src/Diagnostics.DataProviders/K8SELogAnalyticsClient.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            _requestId = RequestId;
+        }
+
+        public async Task<DataTable> ExecuteQueryAsync(string query)
+        {
+            var timeTakenStopWatch = new Stopwatch();
+            QueryResults queryResults;
+            Exception dataProviderException = null;
+            //check query
+
```
Comment (rekhaswaminathan) [inline src/Diagnostics.DataProviders/K8SELogAnalyticsClient.cs:38]: You can add some validation for `query` parameter here.

File: src/Diagnostics.DataProviders/TokenService/K8SELogAnalyticsTokenService.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+        public void Initialize(K8SELogAnalyticsDataProviderConfiguration k8SELogAnalyticsDataProviderConfiguration)
+        {
+            workspaceId = k8SELogAnalyticsDataProviderConfiguration.WorkspaceId;
+            clientId = k8SELogAnalyticsDataProviderConfiguration.ClientId;
+            clientSecret = k8SELogAnalyticsDataProviderConfiguration.ClientSecret;
+
+            domain = k8SELogAnalyticsDataProviderConfiguration.Domain;
+            authEndpoint = k8SELogAnalyticsDataProviderConfiguration.AuthEndpoint;
+            tokenAudience = k8SELogAnalyticsDataProviderConfiguration.TokenAudience;
```
Comment (hforeste) [inline src/Diagnostics.DataProviders/TokenService/K8SELogAnalyticsTokenService.cs:42]: What do partners need to provide so that they can query their log analytics?

File: src/Diagnostics.RuntimeHost/Diagnostics.RuntimeHost.csproj (hunk 1 above)
Comment (DarrelDonald) [reply src/Diagnostics.RuntimeHost/Diagnostics.RuntimeHost.csproj:4]: Yes. The packages needed to use the log analytics api require 2.1. The build fails with 2.2

File: src/Diagnostics.DataProviders/TokenService/K8SELogAnalyticsTokenService.cs (hunk 4 above)
Comment (DarrelDonald) [reply src/Diagnostics.DataProviders/TokenService/K8SELogAnalyticsTokenService.cs:42]: It only queries a specific workspace. To add other workspaces we would need the workspace ID, the domain it's hosted on, and a client ID and secret for an AAD app with access to the workspace.

File: src/Diagnostics.RuntimeHost/Diagnostics.RuntimeHost.csproj (hunk 1 above)
Comment (rekhaswaminathan) [reply src/Diagnostics.RuntimeHost/Diagnostics.RuntimeHost.csproj:4]: Is there any workaround? Not sure what the impact of 2.1 would be and moreover we are trying to progress to latest versions (3.x).

Comment (ShekharGupta1988) [reply src/Diagnostics.RuntimeHost/Diagnostics.RuntimeHost.csproj:4]: I also had to do this change on my local proj to get the runtimehost working. @DarrelDonald - lets not push this change for now and keep it only on your box.

File: src/Diagnostics.DataProviders/K8SELogAnalyticsClient.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        //convert results to DataTable
+        private DataTable ResultAsDataTable(QueryResults results)
+        {
+            DataTable dataTable = new DataTable("results");
+            dataTable.Clear();
+
+            int cursor = 0;
+            foreach (var i in results.Tables[0].Columns)
```
Comment (ShekharGupta1988) [inline src/Diagnostics.DataProviders/K8SELogAnalyticsClient.cs:86]: why do you need cursor variable here. You can simply use i variable in foreach to access the columns[cursor].
Regardless, explore System.Linq in c# to achieve these things in a much more smaller and compact code.
https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.selectmany?view=net-5.0

Comment (ShekharGupta1988) [reply src/Diagnostics.DataProviders/K8SELogAnalyticsClient.cs:86]: Also, since the structures of Data Table and results.Table[0] seems very similar, see if you can simply typecast the object (if that works)

File: src/Diagnostics.DataProviders/K8SELogAnalyticsClient.cs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+namespace Diagnostics.DataProviders
+{
+    class K8SELogAnalyticsClient : ILogAnalyticsClient
+    {
+        private Microsoft.Rest.ServiceClientCredentials creds;
+        private string workspaceId;
+        private static OperationalInsightsDataClient client;
+        private string _requestId;
+
+        public K8SELogAnalyticsClient(string RequestId)
```
Comment (ShekharGupta1988) [inline src/Diagnostics.DataProviders/K8SELogAnalyticsClient.cs:23]: nitpick: small r in requestId parameter

File: src/Diagnostics.DataProviders/LogAnalyticsClientBase.cs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            finally
+            {
+                timeTakenStopWatch.Stop();
+                var StopTime = DateTime.Now;
+                var StartTime = StopTime.AddMilliseconds(-timeTakenStopWatch.ElapsedMilliseconds);
+
+                if (dataProviderException != null)
+                {
+                    DiagnosticsETWProvider.Instance.LogDataProviderException(_requestId,
+                        "K8SELogAnalyticsDataProvider", StartTime.ToString("HH:mm:ss.fff"),
```
Comment (ShekharGupta1988) [inline src/Diagnostics.DataProviders/LogAnalyticsClientBase.cs:45]: move the data provider name - "K8SELogAnalyticsDataProvider" as well to the derived class.

File: src/Diagnostics.DataProviders/LogAnalyticsClientBase.cs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                }
+
+            }
+
+            if (queryResults == null)
+            {
+                queryResults = new QueryResults();
+            }
+            var dataTable = ResultAsDataTable(queryResults);
+            //DataTable dataTable = Enumerable.Cast<DataTable>((Enumerable) queryResults.Tables[0]);
```
Comment (ShekharGupta1988) [inline src/Diagnostics.DataProviders/LogAnalyticsClientBase.cs:63]: remove this comment

**Review by ShekharGupta1988 — APPROVED** [review]
Left few small comments. Rest of the code looks good to me.
Great work in abstracting out the common logic to base classes :)
